### PR TITLE
feat: T2S-paper-download-agent-for-PubMed

### LINKED ISSUE #173: Extend Paper Download Agent to Support PubMed, BioRxiv, MedRxiv, and Semantic Scholar Open Access PDFs
**Is your feature request related to a problem? Please describe.**  
Currently, in Talk2Scholars, the paper download agent only supports fetching PDFs from arXiv using an arXiv ID as input. However, users working in life sciences often need papers from other major repositories such as PubMed, BioRxiv, MedRxiv, and Semantic Scholar. Not supporting these sources limits the accessibility and reach of the system, especially for open-access literature in life sciences.

**Describe the solution you'd like**  
Extend the existing paper download agent to also fetch PDFs from:

- **PubMed API**: Utilize the Entrez Programming Utilities (E-utilities) to access PubMed records.  
  Documentation: [https://www.ncbi.nlm.nih.gov/home/develop/api/](https://www.ncbi.nlm.nih.gov/home/develop/api/)

- **BioRxiv API**: Access preprint metadata and PDFs via the bioRxiv API.  
  Documentation: [https://api.biorxiv.org/](https://api.biorxiv.org/)

- **MedRxiv API**: Similar to bioRxiv, utilize the medRxiv API for accessing preprint data.  
  Documentation: [https://api.medrxiv.org/](https://api.medrxiv.org/)

- **Semantic Scholar API**: Leverage the Semantic Scholar Academic Graph API to retrieve paper data and available open-access PDFs.  
  Documentation: [https://www.semanticscholar.org/product/api](https://www.semanticscholar.org/product/api)

This would work similarly to the arXiv download — by using the respective identifiers (e.g., PubMed ID, DOI, etc.) to retrieve PDFs or fallback links when available.

**Describe alternatives you've considered**  
- Maintaining a completely separate agent for each source (but it's more efficient to unify under one expanded paper download agent).

- Only using Semantic Scholar as a primary search layer to find PubMed/BioRxiv/MedRxiv papers if they are open-access (but direct API integration for each vendor will ensure higher success rates and better control).

**Additional context**  
- Semantic Scholar is already tightly integrated into the Talk2Scholars system through the `s2_agent`, making it feasible to extend it for PDF retrieval as well.

- Focus on open-access and freely available PDFs to avoid licensing issues.

- Consider using DOIs and fallback logic if multiple sources are available for a given paper.

### PR DESCRIPTION
**Pull request by varsha1206**
# For authors

## Description
I have introduced a new platform for downloading research papers. The research papers can now be downloaded from PubMed Central using the PMC ID or PM ID. This enhancement allows for more comprehensive access to research papers, especially those hosted on PubMed Central.

In order to integrate this functionality, I’ve modified the existing code to extract the PMC ID (PubMed Central Identifier) and pass it to the API to fetch paper metadata and download links. The tool successfully retrieves articles when provided with these ids; PM IDs <1234567> and PMC IDS <PMC123456>.

This change requires a modification to the existing S2 system to ensure the PMC ID is extracted and processed correctly.

# Demo Video
https://github.com/user-attachments/assets/81708fa2-0daf-4160-9ec0-7ff07395ded1

## Note
In my next PR, the following issues that you might encounter with this current PR will be solved:
- The download agent will be able to go and download papers from the right source without explicit mention of the source in the prompt. [Asks the user for the source if none provided, in case it doesn't find the right source]
- Abstract class for all the future and existing download functions will be provided to maintain uniformity. 
- The Download Agent will be modularised with utility functions and a single universal tool that directs to the right source.

You can take a look at the work in progress here for more clarity: 
[Download Agent Tools Modifications](https://github.com/varsha1206/AIAgents4Pharma/tree/feat/download-paper-tools-abstraction)

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Congratulations on your first PR 🎉! The code appears well documented and was easy to review. The tests looked very nice which I was very happy to see 🤩. I had mostly minor comments on naming.

File: aiagents4pharma/talk2scholars/tools/paper_download/download_pubmed_paper.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            },
+        timeout=10
+        )
+    response.raise_for_status()
+    return ET.fromstring(response.text)
+
+def extract_metadata(root: ET.Element, pmc_id: str, pdf_download_url: str) -> dict:
+    """Extract metadata for the PMC ID."""
+    title_elem = root.find('.//article-title')
+    title = title_elem.text if title_elem is not None else "N/A"
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/tools/paper_download/download_pubmed_paper.py:74]: I would recommend just leaving this as a blank string. This would be easier to check as opposed to having to know to check for N/A

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Updates look good 🙏. Only a small comment from last time.

File: aiagents4pharma/talk2scholars/tools/paper_download/download_pubmed_paper.py (hunk 1 above)
Comment (varsha1206) [reply aiagents4pharma/talk2scholars/tools/paper_download/download_pubmed_paper.py]: Hey absolutely! I've made the changes as requested :) Please let me know about further updates :)

File: aiagents4pharma/talk2scholars/tools/paper_download/download_pubmed_paper.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    # Publication Data
+    pub_date_elem = root.find('.//published')
+    pub_date = pub_date_elem.text.strip() if pub_date_elem is not None else ""
+
+    pdf_url = f"{pdf_download_url}{pmc_id}?pdf=render"
+    if requests.get(pdf_url,timeout=10).status_code != 200:
+        raise RuntimeError(f"No PDF found or access denied at {pdf_url}")
+
+    logger.info("Metadata from PubMedx for paper PMC ID: %s", pmc_id)
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/tools/paper_download/download_pubmed_paper.py:95]: PubMed

File: aiagents4pharma/talk2scholars/tests/test_paper_download_tools_pubmed.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        # PMC ID
+        pmc_id = "PMC123456"
+
+        # Dummy XML with required fields
+        dummy_xml = """
+        <article>
+            <front>
+                <article-meta>
+                    <title-group>
+                        <article-title>Sample PubMedX Title</article-title>
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/tests/test_paper_download_tools_pubmed.py:43]: PubMed

File: aiagents4pharma/talk2scholars/tests/test_paper_download_tools_pubmed.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+        tool_call_id = "test_tool_id"
+        tool_input = {"pmc_id": pmc_id, "tool_call_id": tool_call_id}
+        result = download_pubmed_paper.run(tool_input)
+        update = result.update
+
+        self.assertIn("article_data", update)
+        self.assertIn(pmc_id, update["article_data"])
+        metadata = update["article_data"][pmc_id]
+        self.assertEqual(metadata["Title"], "Sample PubMedX Title")
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/tests/test_paper_download_tools_pubmed.py:78]: PubMed

File: aiagents4pharma/talk2scholars/tests/test_paper_download_tools_pubmed.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        mock_initialize.return_value.__enter__.return_value = None
+
+        pmc_id = "PMC123456"
+
+        dummy_xml = """
+        <article>
+            <front>
+                <article-meta>
+                    <title-group>
+                        <article-title>Sample PubMedX Title</article-title>
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/tests/test_paper_download_tools_pubmed.py:155]: PubMed

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
Hi @varsha1206 

Congratulations on your first on AA4P! 😃

I have just a few minor comments on the code. However, when I tried to run it, unfortunately it didn’t work 😭

I was specifically interested in working with the article [PMID: 31143927](https://pubmed.ncbi.nlm.nih.gov/31143927/). I made two attempts (see the first two images below): in the first, I provided the article ID with the “PMID” prefix, and in the second, I used just the numerical ID. Sadly, both attempts returned incorrect answers.

From the logs (see the 3rd image below), it looks like the article was successfully downloaded and split into chunks, but it wasn’t used for the Q&A as expected. Instead, it always defaults to my Zotero library to answer the question (@ansh-info).

image


image


image

**Comment by gurdeep330** [comment]
> > * e download agent will be able to go and download papers from the right source without explicit mention of the source in the promp
> 
> Hey! I can see that it isnt working as expected for you, however ive changed nothing and this is working for me, please try running it again or please suggest a possible workaround for it image

Hi Varsha,

A quick observation regarding the graph recursion error that sometimes comes up during your tool execution. This typically occurs when an error arises in the code, often because the LLM passes an argument to the tool that it isn't equipped to handle. These issues can be tricky to trace, especially in LangGraph, since the error is raised internally, and the LLM may continue retrying with the same input until it eventually receives a response (@ansh-info may have his own horror stories to share 😃).

The LLM can adapt—for instance, if it realises that a certain format is required (e.g., the tool expects a PM ID prefixed with "PMC" but receives only the numeric part). However, if the tool doesn’t provide a clear enough error message, the model might keep retrying with the same incorrect input, causing the tool to repeatedly fail.

It might help to handle these edge cases explicitly—for example, raising a more informative error like "PMC prefix missing" (just an example). This way, the LLM has a better chance of correcting its input in the next iteration.

Aside from addressing the test error (please see the Action), I’d suggest reviewing the tool code once more to ensure we’re accounting for such edge cases. Catching these early on can help make the tool more robust.
